tests/h/services/bulk_executor/_actions_test.py

```diff
@@ -0,0 +1,112 @@
+from copy import deepcopy
+
+import pytest
+from h_matchers import Any
+
+from h.h_api.bulk_api import Report
+from h.h_api.exceptions import ConflictingDataError, UnsupportedOperationError
+from h.models import User
+from h.services.bulk_executor._actions import UserUpsertAction
+from tests.h.services.bulk_executor.conftest import CommandFactory
+
+
+def assert_models_match_data(models_by_name, attrs_by_name, handlers=None):
```

Comment:
`handlers` seems to be unused.

If there's anything that could be done to simplify `assert_models_match_data()` and `assert_users_match_commands()`, making them "dumber", I think that'd probably be a good idea. As they stand I think they're a bit on the complicated side for test code. They're hard to follow, and look like they need unit tests of their own.

If the tests could be easier to understand then it'd be easier to look at them and verify that they correctly test that the SQLAlchemy code is correct.

I think it'd be okay for each test to know what models it expects to find in the DB and to express those expected models explicitly, even if that results in some duplication in the tests.

Each test could do something along the lines of:

```python
assert db_session.query(models.User) == [
    UserModelWith(username="user_1", ...),
    UserModelWith(username="user_2", ...),
    UserModelWith(username="user_3", ...),
]
```

(where `UserModelWith()` is a simple/classic-style matcher-class, i.e. a class with an `__eq__()`, just defined in this test module)

Reply:
I've been after something like this for a while, I'll see if I can come up with an extension for `Any` that does something like this.